lib/pages/LoginForm.dart

```diff
@@ -0,0 +1,191 @@
+}
+
+// returns box if match, Error if not
+Future<dynamic> _checkPinHash(String pin) async {
+  SharedPreferences prefs = await SharedPreferences.getInstance();
+
+  HashMap values = HashMap();
+  values["pin"] = pin;
+  values["salt"] = prefs.getString("salt")!;
+
+  final HashMap results = await compute(_calculatePinHash, values);
+  final storedHashedUserKey = prefs.getString("hashedUserKey")!;
+
+  if (results["hash"] != storedHashedUserKey) {
+    return Error();
+  }
+
+  values.clear();
+  values["userKey"] = results["userKey"];
+  values["cipherText"] = prefs.getString("randKeyCipherText");
+  values["nonce"] = prefs.getString("randKeyNonce");
+  values["mac"] = prefs.getString("randKeyMAC");
+
+  final String randKeyStr = await compute(_getRandKey, values);
+  final List<int> randKeyBytes = hex.decode(randKeyStr);
+
+  final calendarBox = await Hive.openBox('calendarBox', encryptionCipher: HiveAesCipher(randKeyBytes));
+  return calendarBox;
+}
+
+class LoginForm extends StatefulWidget {
+  final bool isError;
+
+  const LoginForm({Key? key, required this.isError}) : super(key: key);
+
+  @override
+  State<LoginForm> createState() => _LoginFormState();
+}
+
+class _LoginFormState extends State<LoginForm> {
+  final _formKey = GlobalKey<FormState>();
+  final TextEditingController _pin = TextEditingController();
+
+  @override
+  Widget build(BuildContext context) {
+    return Form(
+      key: _formKey,
+      child: Padding(
+        padding: const EdgeInsets.all(10.0),
+        child: Column(
+          mainAxisAlignment: MainAxisAlignment.center,
+          crossAxisAlignment: CrossAxisAlignment.start,
+          children: <Widget>[
+            const Text(
+              "Enter PIN",
+              style: TextStyle(fontWeight: FontWeight.bold),
+            ),
+            TextFormField(
+              controller: _pin,
+              autocorrect: false,
+              enableSuggestions: false,
+              enableIMEPersonalizedLearning: false,
+              obscureText: true,
+              autofocus: true,
+              validator: (value) {
+                if (value == null || value.isEmpty) {
+                  return 'Please enter your PIN';
+                }
+                return null;
+              },
+              textInputAction: TextInputAction.go,
+              onFieldSubmitted: (value) {
+                _submitLogin(_formKey, context, _pin.text);
+              },
+              decoration: InputDecoration(
+                errorText: widget.isError ? "Incorrect PIN. Please try again." : null,
+              ),
+            ),
+            ElevatedButton(
+              onPressed: () {
+                _submitLogin(_formKey, context, _pin.text);
+              },
+              child: const Text('Log In'),
+            )
+          ],
+        ),
+      ),
+    );
+  }
+}
+
+class LoginPage extends StatefulWidget {
+  final bool isError;
```

Comment:
This passing of a boolean through three different classes seems really strange to me, especially given that we pass it through as a final value that doesn't change. I also don't really like having three different classes for the login form. This LogingPage wrapper is easy to remove since it just passes the isError to a single child component. We can get rid of this and just use the LoginForm directly.

It seems like keeping the error state internally in the class that builds the form makes more sense. That way, when the validator fails (which might also be able to be made part of the form class itself rather than a global?) it can just update the local state simply and render whatever error message it leads to.

Reply:
Definitely agree that the separate class for the scaffold is useless. Fixing that. 

I think I get what you're saying about making the error state a local state vs a parameter, but I'm not 100% sure how we'd implement that. I think we could take this discussion offline — I'll shoot you a message on Discord